src/marqo/s2_inference/x_clip_utils.py

```diff
@@ -0,0 +1,282 @@
+                return True
+            else:
+                False
+                # raise ValueError(f"{thing} cannot be identified as a local file, url or video")
+
+    # if it is an array, then it is an video
+    elif isinstance(thing, list) and isinstance(thing[0], ImageType):
+        return True
+    elif isinstance(thing, ImageType):
+        return True
+    else:
+        raise TypeError(f"expected type Video or str for inputs but received type {type(thing)}")
+
+
+class XCLIP:
+    """
+    conveniance class wrapper to make clip work easily for both text and video encoding
+    """
+
+    def __init__(self, model_type: str = "microsoft/xclip-base-patch16-kinetics-600", device: str = 'cpu', embedding_dim: int = None,
+                 truncate: bool = True, **kwargs) -> None:
+
+        self.model_type = model_type
+        self.device = device
+        self.model = None
+        self.tokenizer = None
+        self.processor = None
+        self.embedding_dimension = embedding_dim
+        self.truncate = truncate
+
+    def load(self) -> None:
+
+        # https://huggingface.co/models?other=xclip
+        self.processor = XCLIPProcessor.from_pretrained(self.model_type)
+        self.model = XCLIPModel.from_pretrained(self.model_type)
+        self.tokenizer = CLIPTokenizer.from_pretrained(self.model_type)
+        self.visionmodel = XCLIPVisionModel.from_pretrained(self.model_type)
+        self.num_input_frames = self.visionmodel.config.num_frames
+        self.model.eval()
+
+    def _convert_output(self, output):
+
+        if self.device == 'cpu':
+            return output.numpy()
+        elif self.device.startswith('cuda'):
+            return output.cpu().numpy()
+
+    @staticmethod
+    def normalize(outputs):
+        return outputs.norm(dim=-1, keepdim=True)
+
+    def encode_text(self, sentence: Union[str, List[str]], normalize=True) -> FloatTensor:
+
+        if self.model is None:
+            self.load()
+
+        text = self.tokenizer(sentence, padding = True, return_tensors = "pt").to(self.device)
+
+        with torch.no_grad():
+            outputs = self.model.get_text_features(**text)
+
+        if normalize:
+            _shape_before = outputs.shape
+            outputs /= self.normalize(outputs)
+            assert outputs.shape == _shape_before
+
+        return self._convert_output(outputs)
+
+    def encode_video(self, videos: Union[str, VideoType, ndarray, List[Union[str, ImageType]]],
+                     normalize=True) -> FloatTensor:
+        if self.model is None:
+            self.load()
+        # default to batch encoding
+        # Note that a single video also is a list
+        if isinstance(videos, list) and isinstance(videos[0], list):
+            video_input = format_and_load_XCLIP_videos(videos)
+        elif isinstance(videos, list) and isinstance(videos[0], ImageType):
+            video_input = [format_and_load_XCLIP_video(videos)]
+
+        #print("processing video")
```

Comment:
we can remove these commented statements

Reply:
done